synapseadmin/roomapi.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2023 Tulir Asokan
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package synapseadmin
+
+import (
+	"context"
+	"net/http"
+	"maunium.net/go/mautrix"
+	"maunium.net/go/mautrix/id"
+	"strconv"
+)
+
+type ReqListRoom struct {
+	SearchTerm	string
+	From		int
+	Limit		int
+}
+
+func (req *ReqListRoom) BuildQuery() map[string]string {
+	query := map[string]string{
+		"from": strconv.Itoa(req.From),
+	}
+	if req.SearchTerm != "" {
+		query["search_term"] = req.SearchTerm
+	}
+	if req.Limit != 0 {
+		query["limit"] = strconv.Itoa(req.Limit)
+	}
+	return query
+}
+
+//  Get room info based on room alias name
+//
+// https://matrix-org.github.io/synapse/latest/admin_api/rooms.html
+func (cli *Client) ListRoom(ctx context.Context, req ReqListRoom) (RoomsResponse ,error) {
+	var resp RoomsResponse
+	var reqURL string
+	reqURL = cli.BuildURLWithQuery(mautrix.SynapseAdminURLPath{"v1", "rooms"}, req.BuildQuery())
+	_, err := cli.MakeFullRequest(ctx, mautrix.FullRequest{
+		Method:       http.MethodGet,
+		URL:          reqURL,
+		ResponseJSON: &resp,
+	})
+	return resp, err
+}
+
+// ReqDeleteRoom is the request content for Client.DeleteRoom.
+type ReqDeleteRoom struct {
+	// Default to true: remove all traces from database
+	Purge bool `json:"purge"`
```

Comment:
This should also have the other fields (`block`, `message`, `room_name`, `new_room_user_id`). Also, all the fields are optional, so they should have `omitempty` in the json struct tag